airbyte-integrations/connectors/source-relational-db/src/main/java/io/airbyte/integrations/source/relationaldb/AbstractDbSource.java

```diff
@@ -180,6 +183,34 @@ public AutoCloseableIterator<AirbyteMessage> read(final JsonNode config,
         });
   }
 
+  private void validateSourceSchema(Map<String, TableInfo<CommonField<DataType>>> fullyQualifiedTableNameToInfo,
+                                    ConfiguredAirbyteCatalog catalog,
+                                    Database database) {
+    for (final ConfiguredAirbyteStream airbyteStream : catalog.getStreams()) {
+      final AirbyteStream stream = airbyteStream.getStream();
+      final String fullyQualifiedTableName = getFullyQualifiedTableName(stream.getNamespace(),
+          stream.getName());
+      if (!fullyQualifiedTableNameToInfo.containsKey(fullyQualifiedTableName)) {
+        continue;
+      }
+      final TableInfo<CommonField<DataType>> table = fullyQualifiedTableNameToInfo.get(fullyQualifiedTableName);
+      final List<Field> fields = table.getFields()
+          .stream()
+          .map(this::toField)
+          .distinct()
+          .collect(Collectors.toList());
+      final JsonNode currentJsonSchema = fieldsToJsonSchema(fields);
+
+      final JsonNode catalogSchema = stream.getJsonSchema();
+      if (!catalogSchema.equals(currentJsonSchema)) {
+        LOGGER.warn("Source schema changed for table  {}! Actual schema: {}. Catalog schema:  {}",
```

Comment:
Is this an error state?
Should we do something other than warn?

Reply:
+1. Can we actually throw a `ConfigErrorException` here so that this is actionable to the user? With a message maybe : "The underlying schema changed for the table. Please refresh your source schema! Source schema changed for table {}! Actual schema: {}. Catalog schema: {}" 

Along those lines. In addition, could you add a unit test as well?

Reply:
@rodireich @akashkulk 
I'm not sure that throwing an exception here would be correct in all cases, the sync would fail if the datatype changed from integer to float, but if the change were the reverse (from float to integer), and all values in the database contained only integers, then the sync would be successful, also if the user manually added some columns - the sync will be successful, but in some cases when the column is deleted - the sync will fail
I think that there may be other possible variations of a successful/unsuccessful sync, for example, using an incremental mode, but at the same time, throwing an exception here will lead to a complete fail of the sync, while only one stream can fail, for which the user manually changed structure in the database, and all other streams should not fail
WDYT?

Reply:
I see. So some changes can go through but other will require to refresh the source schema.
My question: 
for non-breaking change (e.g user adds a column) - can we keep on syncing indefinitely with the old schema?
for breaking change (e.g user removes a column) - at what stage will the sync fail, is it deterministic or depending on sync type and other factors.

I think this change is making it better for identifying a state we're unaware of today, until something breaks. I wonder if we there's a good way to know what to do? (maybe it's impossible).
+1 to Akash's suggestion for adding a test

Reply:
@rodireich added the unit test

Answering your questions:
for non-breaking change - yes, sync could be indefinitely with the old schema
for breaking change - i'm not sure that we could determine the exact stage of sync failure . Potentially - it could be 1. On normalization step( as described [here]([21413](https://app.zenhub.com/workspaces/db--dw-source-connectors-6333360e0a41155061efbcbd/issues/gh/airbytehq/airbyte/21413))) , 
2. During WRITE(avro values mapping to the old avro schema or during json schema formatting in BigQuery Denormalized), 
3. During READ (at any place where old catalog used)